drivers/watchdog/virtio_wdt.c

```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Watchdog driver for virtio. Derived from virto-rng.c and i6300esb.c
+ * Copyright 2007, 2008 Rusty Russell IBM Corporation
+ * Copyright 2004 Google Inc.
+ * Copyright 2005 David Härdeman <david@2gen.com>
+ * Copyright 2020 Intel Corporation
+ */
+
+#include <linux/err.h>
+#include <linux/scatterlist.h>
+#include <linux/spinlock.h>
+#include <linux/virtio.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/watchdog.h>
+#include <linux/virtio_ids.h>
+#include <linux/virtio_config.h>
+
+static DEFINE_IDA(watchdog_index_ida);
+
+#define to_info(wptr) container_of(wptr, struct virtio_watchdog_info, wdd)
+
+/* Only support an interval of 15s */
+#define VW_HEARTBEAT_DEFAULT 15
+
+struct virtio_watchdog_info {
+	struct watchdog_device wdd;
+	struct virtio_device *vdev;
+	struct virtqueue *vq;
+	struct completion have_data;
+	char name[25];
+	unsigned int data_avail;
+	int index;
+	bool busy;
+	bool wdd_register_done;
+};
+
+static void virtio_watchdog_recv_done(struct virtqueue *vq)
+{
+	struct virtio_watchdog_info *vi = vq->vdev->priv;
+
+	/* We can get spurious callbacks, e.g. shared IRQs + virtio_pci. */
+	if (!virtqueue_get_buf(vi->vq, &vi->data_avail))
+		return;
+
+	complete(&vi->have_data);
+}
+
+/* Host will change the buffer from 0->1 */
+static void register_buffer(struct virtio_watchdog_info *vi, u8 *buf,
+							size_t size)
+{
+	struct scatterlist sg;
+
+	sg_init_one(&sg, buf, size);
+	virtqueue_add_inbuf(vi->vq, &sg, 1, buf, GFP_KERNEL);
+
+	virtqueue_kick(vi->vq);
+}
+
+
+static int virtio_watchdog_ping(struct watchdog_device *wdd)
+{
+	struct virtio_watchdog_info *vi = to_info(wdd);
+	struct virtio_device *vdev = vi->vdev;
+	int ret;
+	u8 *buf;
+
+	if (!vi->wdd_register_done)
+		return -ENODEV;
+
+	buf = kzalloc(sizeof(u8), GFP_KERNEL);
+	if (!vi->busy) {
+		vi->busy = true;
+		reinit_completion(&vi->have_data);
+		register_buffer(vi, (void *)buf, sizeof(u64));
+	}
+
+	ret = wait_for_completion_killable(&vi->have_data);
+	if (ret < 0)
+		goto err;
+
+	if (*buf != 1) {
+		dev_err(&vdev->dev,
+				"Host did not acknowledge buffer correctly");
+		ret = -EINVAL;
+	}
+
+err:
+	vi->busy = false;
+	kfree(buf);
+
+	return ret;
+}
+
+static int virtio_watchdog_start(struct watchdog_device *wdd)
+{
+	struct virtio_watchdog_info *vi = to_info(wdd);
+	struct virtio_device *vdev = vi->vdev;
+
+	dev_info(&vdev->dev, "Watchdog started");
+
+	return 0;
+}
+
+static int virtio_watchdog_stop(struct watchdog_device *wdd)
+{
+	struct virtio_watchdog_info *vi = to_info(wdd);
+	struct virtio_device *vdev = vi->vdev;
+
+	dev_info(&vdev->dev, "Watchdog stop request ignored");
```

Comment:
Since this is a no-op, I think we shouldn't define the callback at all and let the caller (`watchdog_stop()` from `drivers/watchdog/watchdog_dev.c`) deal with it.

Reply:
The spec says that `->stop()` is not required. The code however does need it (I tried deleting it and got `-EINVAL` on registration).

Reply:
Ah fair enough!